_posts/2021-01-04-unit-tests-and-concurrency.md

```diff
@@ -0,0 +1,135 @@
+Using RxJava this looks like:
+
+```kotlin
+// Emits Loading then Content or Problem
+private fun requestData(): Observable<ViewState> =
+    service.firstApiCall()
+        .observeOn(Schedulers.computation())
+        .map { it.message }
+        .flatMap(this::secondApiCall)
+        .map<ViewState> { ViewState.Content(it) }
+        .startWith(Single.just(ViewState.Loading))
+        .onErrorReturn { ViewState.Problem }
+        .toObservable()
+
+// Concurrently executes secondApiCall for each element in list. 
+// Transforms the result to ViewEntity, combines everything in a list
+private fun secondApiCall(list: List<String>): Single<List<ViewEntity>> =
+    Observable.fromIterable(list)
+        .concatMapEager { item ->
+            service.secondEndpoint(item)
+                .observeOn(Schedulers.computation())
+                .map { it.message }
+                .map { ViewEntity(item, it.toHttpUrl()) }
+                .onErrorReturn { DogViewEntity(item, null) }
+                .toObservable()
+        }
+        .toList()
+```
+
+Since the list is 15-20 elements, I am using `concatMapEager` to execute the API calls for the second endpoint concurrently while maintaining the order of the elements. This worked great and I wrote my unit tests using the real schedulers. RxJava has amazing testing support that enables this. Thanks [Simon Vergauwen][simon] for the tips on how to do that.
+
+```kotlin
+@Test
+fun `execute - first call succeeds, one failed call to #2 - one full item, one item with default value`() {
+    val items = listOf("Item 1" to "https://item1.jpg", "Item 2" to "invalid url")
+    val service = successService(items)
+
+    val viewModel = ViewModel(service)
+
+    val expected = listOf(
+        ViewEntity("Item 1", "https://item1.jpg".toHttpUrl()),
+        ViewEntity("Item 2", null)
+    )
+    viewModel.state
+        .test()
+        .also { viewModel.execute() }
+        .awaitCount(2)  // (0) Loading, (1) Content
+        .assertValueAt(1, Content(expected))
+}
+```
+
+Here, `awaitCount(2)` makes sure the assertions after it are ran only after it gets at least 2 items. Again RxJava has excellent testing support.
+
+## The Refactor
+
+This worked reasonably well, however I decided to try and optimise it further. 
+
+My current code would: 
+
+- start with a progress bar
+- do API call #1 
+- do API call #2 N times concurrently 
+- combine the results 
+- show the full list in the UI 
+
+What if we display the data as it becomes available. 
+
+- start with a progress bar 
+- do API call #1 
+- show the data from API call #1 in the UI 
+- do API calls #2 N times concurrently without displaying progress 
+- update the UI N times as the API calls for #2 are completing
+
+I was already using `ListAdapter` which has built-in diffing support. I can send an updated list and it will do it's magic to update the changed elements only.
+
+## First attempt
```

Comment:
Nitpicking: the fact this section is called "First attempt" made me think it would have more attempts but afterwards you move to conclusion. 🤔

Reply:
+1

Reply:
I wanted to also explain my solution.... but I prefer short and to the point blog posts.... also, I am not sure it's the optimal one :see_no_evil: